src/controller_examples/zookeeper_controller/spec/zookeeper_api.rs

```diff
@@ -8,92 +8,208 @@ use vstd::{prelude::*, string::*};
-// by querying the cluster state.
-//
-// TODO: specify version number of zookeeper node to reason about potential race between the current API call and
-// the old API call from the previously crashed controller.
-pub open spec fn set_zk_node(
-    zk_name: StringView, zk_namespace: StringView, replicas: StringView, state: ZooKeeperState
-) -> (ZooKeeperState, Option<ZKAPIOutputView>) {
+// TODO: more validation check could be implemented,
+// such as checking the existence of the service object as well,
+// checking whether the stateful set is really ready,
+// and checking whether the port number is correct (if not hardcoded as it is).
+pub open spec fn validate(name: StringView, namespace: StringView, path: Seq<StringView>, resources: StoredState) -> bool {
     let key = ObjectRef {
-        kind: Kind::CustomResourceKind,
-        namespace: zk_namespace,
-        name: zk_name,
+        kind: Kind::StatefulSetKind,
+        namespace: namespace,
+        name: name,
     };
-    let new_data = state.data.insert(key, replicas);
-    let state_prime = ZooKeeperState {
-        data: new_data,
-        ..state
-    };
-    (state_prime, Some(ZKAPIOutputView::SetZKNodeResponse(ZKAPIResultView{res: Ok(())})))
+    &&& path.len() > 0
+    &&& resources.dom().contains(key)
+}
+
+// handle_exists models the behavior of the zookeeper server handling the exists request.
+// It checks the existence of the node by querying the map.
+// If the node exists, it returns its stat (i.e., version number), otherwise none.
+// Note that it uses the uid to avoid querying the data belonging to the old stateful set object.
+pub open spec fn handle_exists(
+    name: StringView, namespace: StringView, path: Seq<StringView>, resources: StoredState, state: ZKState
+) -> (ZKState, ZKAPIExistsResultView) {
+    let key = ObjectRef { kind: Kind::StatefulSetKind, namespace: namespace, name: name };
+    if !validate(name, namespace, path, resources) {
+        (state, ZKAPIExistsResultView{res: Err(Error::ZKNodeExistsFailed)})
+    } else {
+        let current_uid = resources[key].metadata.uid.get_Some_0();
+        let addr = ZKNodeAddr::new(name, namespace, current_uid, path);
+        if !state.data.dom().contains(addr) {
+            (state, ZKAPIExistsResultView{res: Ok(None)})
+        } else {
+            let version = state.data[addr].1;
+            (state, ZKAPIExistsResultView{res: Ok(Some(version))})
+        }
+    }
+}
+
+// handle_create models the behavior of the zookeeper server handling the create request.
+// The creation succeeds only when (1) the node does not exist yet and (2) the parent node exists.
+pub open spec fn handle_create(
+    name: StringView, namespace: StringView, path: Seq<StringView>, data: ZKNodeValue, resources: StoredState, state: ZKState
+) -> (ZKState, ZKAPICreateResultView) {
+    let key = ObjectRef { kind: Kind::StatefulSetKind, namespace: namespace, name: name };
+    if !validate(name, namespace, path, resources) {
```

Comment:
Actually I am a bit confused. Is it true that when we are here, the node must exist?

Reply:
Which node must exist? You mean the parent node?